Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx

```diff
@@ -1,24 +1,225 @@
+import { stat } from "node:fs";
+import React, { useCallback, useMemo, useReducer, useState } from "react";
+
 import BatteryInput from "~/components/batteryInput";
 import Header from "~/components/header";
 import SpeedInput from "~/components/speedInput";
 import WeatherInput from "~/components/weatherInput";
 
-const App = () => {
-  const calculateRange = () => {
-    return;
+// 1. Define the State interface
+interface State {
+  speed: number | string;
+  battery: number | string;
+  weather: number | string;
+  valid: boolean;
+  changed: {
+    speed: boolean;
+    battery: boolean;
+  };
+}
+
+type ActionType =
+  | { type: "Update speed"; payload: number | string }
+  | { type: "Update battery"; payload: number | string }
+  | { type: "Update weather"; payload: number | string }
+  | { type: "Update valid"; payload: boolean }
+  | { type: "changed"; field: "speed" | "battery"; payload: boolean };
+
+const initialState: State = {
+  speed: "",
+  battery: "",
+  weather: 0,
+  valid: true,
+  changed: {
+    speed: false,
+    battery: false,
+  },
+};
+
+const reducer = (state: State, action: ActionType): State => {
+  switch (action.type) {
+    case "Update speed":
+      return { ...state, speed: action.payload };
+    case "Update battery":
+      return { ...state, battery: action.payload };
+    case "Update weather":
+      return { ...state, weather: action.payload };
+    case "Update valid":
+      return { ...state, valid: action.payload };
+    case "changed":
+      return {
+        ...state,
+        changed: { ...state.changed, [action.field]: action.payload },
+      };
+    default:
+      return state;
   }
+};
+const App = () => {
+  const [state, dispatch] = useReducer(reducer, initialState);
+  /*const [speed, setSpeed] = useState<number | string>("");
+  const [battery, setBattery] = useState<number | string>("");
+  const [weather, setWeather] = useState<number | string>(0);*/
+  const { speed, battery, weather } = state; //initialize the state
+  //const [valid, setValid] = useState(true);
+  let range = 0;
+
+  //to avoid the type errors:
+  const handleSpeedChange = useCallback(
+    (e: React.ChangeEvent<HTMLInputElement>) => {
+      const value = e.target.value;
+      dispatch({
+        type: "Update speed",
+        payload: value === "" ? "" : Number(value),
+      });
+      //setSpeed(value === "" ? "" : Number(value));
+      // Dispatch to update touched state when the input changes
+      dispatch({
+        type: "changed",
+        field: "speed", // Correctly specify the field
+        payload: value === "", // Set to true if the value is an empty string
+      });
```

Comment:
calling dispatch back-to-back may cause some undefined behaviour, it may also causes 2 re-renders